### ISSUE TITLE
How to make draw calls from an other thread ?

### ISSUE THREAD

**Issue by dhiahassen** [grimfang4/sdl-gpu#95]
I am really sorry for posting many issue , but i am building a game engin with and IDE .... i am really glad for using sdl_gpu for the game engine , it is very funtional and it exports on all platforms sdl supports ....

In the engine studio the game is loaded as dll in a seperate thread , when the game performs calls using my api from it's thread , the calls are all batched and then sent to the gpu from the studio's main thread when the display is about to be fliped , i have done a search and found here that i can switch opengl context from one thread to another , i saw that you are using such thing as "GPU_SetCurrent" or something ( glMake Current() is used for the pupsose )  ... so how can i draw from a seperate thrread using SDL_gpu without batching all calls ?

**Comment by grimfang4** [comment]
Hey, no problem at all.  This is a great place to ask questions.

So it sounds like you're doing what I would do for multiple threads: Create a separate GL context for each thread, one thread per window.  One context per thread can avoid serialized vsync issues for multiple windows (a problem I've seen on Linux).  You do need a window as the basis for a GL context.

Have you tried making these shared contexts?  I use `SDL_GL_SetAttribute(SDL_GL_SHARE_WITH_CURRENT_CONTEXT, 1);` before GPU_CreateTargetFromWindow() to make contexts able to use some of the same GL objects.  Textures are shared, for example, but FBOs are not.  You should be able to blit a texture from a different context onto the current context's render targets, but you will probably have trouble trying to do it the other way around.

**Comment by dhiahassen** [comment]
Yes , i forgot to mention that i tried  'SDL_GL_SetAttribute(SDL_GL_SHARE_WITH_CURRENT_CONTEXT, 1);' and i still get the gl_handle = 0 for game_target->image->data->handle  , that means opengl didnt share the textures , in other words , opengl started couting again from 0 for texture ids , thats no supposed to happen i think, because both contexts are new supposed to share the id(s counter for textures , and blting does not work , sdl gpu crashes when trying to blit an image made in context A to a target in context B , if only the is a solution for that , everything else is solved

**Comment by grimfang4** [comment]
I haven't gotten around to testing this on multiple threads, but I'll get to it.

**Comment by Dids** [comment]
Any news on this?

I'm attempting something similar, where I need to capture frames on a separate thread, which from what I can tell is only possible by creating a separate target on the separate thread.

Enabling shared contexts and creating the target in the thread itself results in an exception in `glGenTextures(1, &handle)`, among other places.

I'm simply passing the current window handle to the thread, then creating a new target from that, but I'm assuming that's the right way to do it anyway, and the issue is elsewhere.

That said, is there any other way of capturing frames (eg. screen->image) without blocking the main thread while doing so?

**Comment by dhiahassen** [comment]
what i am doing is using a blit batch , i created a wrapper class for GPU_Target , then that target can be attached to an object of type 'BlitBatch' , and when a draw call gets performed on that target , if batching is 'ON' then no direct draw is made , instead of that , the draw call will be appended to the  blit batch of that target , and later when the child thread is done with all draw calls of this frame , he he does this : 

` ` 
 ....
BEGIN_DRAW_CALLS()
     target->GetBlitBatch()->Flush();
END_DRAW_CALLS()
 ....
` ` 

the flush method does finally the actual draw calls but from the parent thread !
how that is done ?
the macros BEGIN_DRAW_CALLS() / END_DRAW_CALLS() wraps the draw calls  and push them in a TaskList of the main thread , thread task list is constantly checked from the main thread each frame , also BEGIN_DRAW_CALLS() should stop the child thread until the tasklist is served , thefore you will need a mutex/bolean .... , i am trying to explain a complex mechanism of 700 lines of code so excuse me ..
Note : you may face a synchronazation problem when bliting one target to another and the draw are not performed in thier actual order , so you can use a global draw batch for the whole engine

**Comment by grimfang4** [comment]
Also, the best way I can think of to offload frame capture to another thread is only a "sort of" solution.  You can't ask another thread to read the framebuffer without interfering with the main thread, but you can ask it to save data to disk.

Render your whole game to an intermediate buffer (which then is normally rendered directly to the screen target).  Then, when doing a screenshot, you'd GPU_CopySurfaceFrom|Image/Target|(intermediate_screen) and pass the resulting surface to the worker thread.  The worker thread can do what it needs to with the surface and free it when done.

**Comment by dhiahassen** [comment]
@grimfang4  you dont understand , any ANY draw call performed from a child thread will cause the app to crash , it is not possible

**Comment by Dids** [comment]
@grimfang4 That's definitely a good idea, and it's more or less what I'm doing right now.

However, I'm also trying to record multiple frames (say 5 seconds at 60 frames per second), which I then turn into a recording.
Simply calling `GPU_CopySurfaceFromTarget` _every_ frame seems like it's killing performance (unless it's something else I'm doing to block the main thread). Any tips on that particular scenario?

**Comment by grimfang4** [comment]
That's good news.  Reading from the framebuffer on the GPU is going to be relatively slow, one way or another, but it sounds like it's fast enough for that purpose.

**Comment by dhiahassen** [comment]
@Dids  a new vulkan version is in process , so this can be fixed i think because i heard that vulkan is much more multithread friendly
